File: flutter_implementation/lib/features/reader/tajweed/tajweed_engine.dart
import '../../../core/models/tajweed_rule.dart';

/// Engine for detecting Tajweed rules in Arabic Quranic text
class TajweedEngine {
  /// Analyze Arabic text and return all Tajweed spans
  List<TajweedSpan> analyze(String arabicText) {
    if (arabicText.isEmpty) return [];

    List<TajweedSpan> spans = [];

    // Order matters: Check specific rules first, then general ones
    spans.addAll(_detectGhunnah(arabicText));
    spans.addAll(_detectIqlab(arabicText));
    spans.addAll(_detectIdghaam(arabicText));
    spans.addAll(_detectIkhfa(arabicText));
    spans.addAll(_detectQalqalah(arabicText));
    spans.addAll(_detectMadd(arabicText));

    // Remove overlaps - prioritize more specific rules
    return _removeOverlaps(spans);
  }

  /// Detect Ghunnah (غُنَّة) - Noon or Meem with Shaddah
  List<TajweedSpan> _detectGhunnah(String text) {
    List<TajweedSpan> spans = [];

    // Pattern: ن or م followed by Shaddah (ّ)
    final pattern = RegExp(r'[نم]\u0651');

    for (final match in pattern.allMatches(text)) {
      spans.add(TajweedSpan(
        start: match.start,
        end: match.end,
        rule: TajweedRule.ghunnah,
        matchedText: text.substring(match.start, match.end),
      ));
    }

    return spans;
  }

  /// Detect Iqlab (إقلاب) - Noon Sakinah/Tanween + ب
  List<TajweedSpan> _detectIqlab(String text) {
    List<TajweedSpan> spans = [];

    // Pattern: Noon Sakinah (نْ) or Tanween (ً ٌ ٍ) followed by ب
    final patterns = [
      RegExp(r'ن\u0652[\s]*ب'), // Noon Sakinah + Ba
      RegExp(r'[\u064B\u064C\u064D][\s]*ب'), // Tanween + Ba
    ];

    for (final pattern in patterns) {
      for (final match in pattern.allMatches(text)) {
        spans.add(TajweedSpan(
          start: match.start,
          end: match.end,
          rule: TajweedRule.iqlab,
          matchedText: text.substring(match.start, match.end),
        ));
      }
    }

    return spans;
  }

  /// Detect Idghaam (إدغام) - Noon Sakinah/Tanween + يرملون
  List<TajweedSpan> _detectIdghaam(String text) {
    List<TajweedSpan> spans = [];

    // Idghaam letters: ي ر م ل و ن
    const idghaamLetters = 'يرملون';

    // Pattern: Noon Sakinah or Tanween + Idghaam letter
    final patterns = [
      RegExp(r'ن\u0652[\s]*[' + idghaamLetters + r']'),
      RegExp(r'[\u064B\u064C\u064D][\s]*[' + idghaamLetters + r']'),
    ];

    for (final pattern in patterns) {
      for (final match in pattern.allMatches(text)) {
        spans.add(TajweedSpan(
          start: match.start,
          end: match.end,
          rule: TajweedRule.idghaam,
          matchedText: text.substring(match.start, match.end),
        ));
      }
    }

    return spans;
  }

  /// Detect Ikhfa (إخفاء) - Noon Sakinah/Tanween + 15 letters
  List<TajweedSpan> _detectIkhfa(String text) {
    List<TajweedSpan> spans = [];

    // 15 Ikhfa letters: ت ث ج د ذ ز س ش ص ض ط ظ ف ق ك
    const ikhfaLetters = 'تثجدذزسشصضطظفقك';

    // Pattern: Noon Sakinah or Tanween + Ikhfa letter
    final patterns = [
      RegExp(r'ن\u0652[\s]*[' + ikhfaLetters + r']'),
      RegExp(r'[\u064B\u064C\u064D][\s]*[' + ikhfaLetters + r']'),
    ];

    for (final pattern in patterns) {
      for (final match in pattern.allMatches(text)) {
        spans.add(TajweedSpan(
          start: match.start,
          end: match.end,
          rule: TajweedRule.ikhfa,
          matchedText: text.substring(match.start, match.end),
        ));
      }
    }

    return spans;
  }

  /// Detect Qalqalah (قلقلة) - قطبجد with Sukoon
  List<TajweedSpan> _detectQalqalah(String text) {
    List<TajweedSpan> spans = [];

    // 5 Qalqalah letters: ق ط ب ج د
    const qalqalahLetters = 'قطبجد';

    // Pattern: Qalqalah letter with Sukoon (ْ) or at the end of a word
    final pattern = RegExp(r'[' + qalqalahLetters + r']\u0652');

    for (final match in pattern.allMatches(text)) {
      spans.add(TajweedSpan(
        start: match.start,
        end: match.end,
        rule: TajweedRule.qalqalah,
        matchedText: text.substring(match.start, match.end),
      ));
    }

    // Also check for Qalqalah letters at word boundaries (common in pauses)
    final wordEndPattern = RegExp(r'[' + qalqalahLetters + r'](?=\s|$)');
    for (final match in wordEndPattern.allMatches(text)) {
      // Only add if not already detected with Sukoon
      final overlaps = spans.any((s) => s.start == match.start);
      if (!overlaps) {
        spans.add(TajweedSpan(
          start: match.start,
          end: match.end,
          rule: TajweedRule.qalqalah,
          matchedText: text.substring(match.start, match.end),
        ));
      }
    }

    return spans;
  }

  /// Detect Madd (مَدّ) - Elongation
  List<TajweedSpan> _detectMadd(String text) {
    List<TajweedSpan> spans = [];

    // Common Madd patterns:
    // 1. Alif with Madd sign (آ)
    // 2. Waw with Sukoon after Dammah (وْ)
    // 3. Ya with Sukoon after Kasrah (يْ)
    // 4. Letters with Madd Lazim (connected or separated)

    final patterns = [
      // Madd with Alif Madd (آ)
      RegExp(r'[\u0622]'), // آ

      // Natural Madd with elongated vowels
      RegExp(r'[اوي]\u0653'), // With Madd sign

      // Madd Munfasil (separated) - vowel + Hamzah
      RegExp(r'[اوي][\s]*[\u0621\u0623\u0624\u0625\u0626]'),

      // Common Madd patterns in specific words
      RegExp(r'ا{2,}'), // Multiple Alifs (rare but possible)
    ];

    for (final pattern in patterns) {
      for (final match in pattern.allMatches(text)) {
        spans.add(TajweedSpan(
          start: match.start,
          end: match.end,
          rule: TajweedRule.madd,
          matchedText: text.substring(match.start, match.end),
        ));
      }
    }

    return spans;
  }

  /// Remove overlapping spans - keep the one with higher priority
  List<TajweedSpan> _removeOverlaps(List<TajweedSpan> spans) {
    if (spans.isEmpty) return spans;

    // Sort by start position
    spans.sort((a, b) => a.start.compareTo(b.start));

    List<TajweedSpan> result = [];
    TajweedSpan? current;

    for (final span in spans) {
      if (current == null) {
        current = span;
        continue;
      }

      // Check for overlap
      if (span.start < current.end) {
        // Overlapping - keep the more specific/important rule
        // Priority: Iqlab > Idghaam > Ghunnah > Ikhfa > Qalqalah > Madd
        final priorities = {
          TajweedRule.iqlab: 6,
          TajweedRule.idghaam: 5,
          TajweedRule.ghunnah: 4,
          TajweedRule.ikhfa: 3,
          TajweedRule.qalqalah: 2,
          TajweedRule.madd: 1,
        };

        final currentPriority = priorities[current.rule] ?? 0;
        final spanPriority = priorities[span.rule] ?? 0;

        if (spanPriority > currentPriority) {
          current = span;
        }
      } else {
        // No overlap - add current and move to next
        result.add(current);
        current = span;
      }
    }

    if (current != null) {
      result.add(current);
    }

    return result;
  }

  /// Get human-readable description of what was detected
  String describeSpan(TajweedSpan span) {
    return '${span.rule.englishName} at position ${span.start}: "${span.matchedText}"';
  }

  /// Get statistics about Tajweed usage in text
  Map<TajweedRule, int> getStatistics(String text) {
    final spans = analyze(text);
    final stats = <TajweedRule, int>{};

    for (final rule in TajweedRule.values) {
      stats[rule] = 0;
    }

    for (final span in spans) {
      stats[span.rule] = (stats[span.rule] ?? 0) + 1;
    }

    return stats;
  }
}

File: client/src/lib/tajweed-engine.ts
import { TajweedRule, TajweedSpan } from './tajweed-rules';

// Arabic diacritical marks (Unicode constants)
const SHADDAH = '\u0651';      // ّ - Emphasis/doubling mark
const SUKOON = '\u0652';       // ْ - No vowel mark
const TANWEEN_FATHATAN = '\u064B';  // ً - Tanween with Fatha
const TANWEEN_DAMMATAN = '\u064C';  // ٌ - Tanween with Damma  
const TANWEEN_KASRATAN = '\u064D';  // ٍ - Tanween with Kasra
const MADD_SIGN = '\u0653';    // ٓ - Madd above
const ALIF_MADD = '\u0622';    // آ - Alif with Madd

// Hamza variants
const HAMZA_VARIANTS = '[\u0621\u0623\u0624\u0625\u0626]';

/**
 * Engine for detecting Tajweed rules in Arabic Quranic text
 */
export class TajweedEngine {
  /**
   * Analyze Arabic text and return all Tajweed spans
   */
  analyze(arabicText: string): TajweedSpan[] {
    if (!arabicText) return [];

    const spans: TajweedSpan[] = [];

    // Order matters: Check specific rules first
    spans.push(...this.detectGhunnah(arabicText));
    spans.push(...this.detectIqlab(arabicText));
    spans.push(...this.detectIdghaam(arabicText));
    spans.push(...this.detectIkhfa(arabicText));
    spans.push(...this.detectQalqalah(arabicText));
    spans.push(...this.detectMadd(arabicText));

    return this.removeOverlaps(spans);
  }

  /**
   * Detect Ghunnah (غُنَّة) - Noon or Meem with Shaddah
   */
  private detectGhunnah(text: string): TajweedSpan[] {
    const spans: TajweedSpan[] = [];
    // Pattern: ن or م followed by Shaddah (ّ)
    const pattern = new RegExp(`[نم]${SHADDAH}`, 'g');
    let match;

    while ((match = pattern.exec(text)) !== null) {
      spans.push({
        start: match.index,
        end: match.index + match[0].length,
        rule: TajweedRule.GHUNNAH,
        text: match[0],
      });
    }

    return spans;
  }

  /**
   * Detect Iqlab (إقلاب) - Noon Sakinah/Tanween + ب
   */
  private detectIqlab(text: string): TajweedSpan[] {
    const spans: TajweedSpan[] = [];

    // Noon Sakinah (نْ) or Tanween (ً ٌ ٍ) followed by ب
    const patterns = [
      new RegExp(`ن${SUKOON}[\\s]*ب`, 'g'),           // Noon Sakinah + Ba
      new RegExp(`[${TANWEEN_FATHATAN}${TANWEEN_DAMMATAN}${TANWEEN_KASRATAN}][\\s]*ب`, 'g'), // Tanween + Ba
    ];

    patterns.forEach(pattern => {
      let match;
      while ((match = pattern.exec(text)) !== null) {
        spans.push({
          start: match.index,
          end: match.index + match[0].length,
          rule: TajweedRule.IQLAB,
          text: match[0],
        });
      }
    });

    return spans;
  }

  /**
   * Detect Idghaam (إدغام) - Noon Sakinah/Tanween + يرملون
   */
  private detectIdghaam(text: string): TajweedSpan[] {
    const spans: TajweedSpan[] = [];
    const idghaamLetters = 'يرملون';

    const patterns = [
      new RegExp(`ن${SUKOON}[\\s]*[${idghaamLetters}]`, 'g'),
      new RegExp(`[${TANWEEN_FATHATAN}${TANWEEN_DAMMATAN}${TANWEEN_KASRATAN}][\\s]*[${idghaamLetters}]`, 'g'),
    ];

    patterns.forEach(pattern => {
      let match;
      while ((match = pattern.exec(text)) !== null) {
        spans.push({
          start: match.index,
          end: match.index + match[0].length,
          rule: TajweedRule.IDGHAAM,
          text: match[0],
        });
      }
    });

    return spans;
  }

  /**
   * Detect Ikhfa (إخفاء) - Noon Sakinah/Tanween + 15 letters
   */
  private detectIkhfa(text: string): TajweedSpan[] {
    const spans: TajweedSpan[] = [];
    const ikhfaLetters = 'تثجدذزسشصضطظفقك';

    const patterns = [
      new RegExp(`ن${SUKOON}[\\s]*[${ikhfaLetters}]`, 'g'),
      new RegExp(`[${TANWEEN_FATHATAN}${TANWEEN_DAMMATAN}${TANWEEN_KASRATAN}][\\s]*[${ikhfaLetters}]`, 'g'),
    ];

    patterns.forEach(pattern => {
      let match;
      while ((match = pattern.exec(text)) !== null) {
        spans.push({
          start: match.index,
          end: match.index + match[0].length,
          rule: TajweedRule.IKHFA,
          text: match[0],
        });
      }
    });

    return spans;
  }

  /**
   * Detect Qalqalah (قلقلة) - قطبجد with Sukoon
   */
  private detectQalqalah(text: string): TajweedSpan[] {
    const spans: TajweedSpan[] = [];
    const qalqalahLetters = 'قطبجد';

    // With Sukoon
    const pattern = new RegExp(`[${qalqalahLetters}]${SUKOON}`, 'g');
    let match;

    while ((match = pattern.exec(text)) !== null) {
      spans.push({
        start: match.index,
        end: match.index + match[0].length,
        rule: TajweedRule.QALQALAH,
        text: match[0],
      });
    }

    // At word end
    const wordEndPattern = new RegExp(`[${qalqalahLetters}](?=\\s|$)`, 'g');
    while ((match = wordEndPattern.exec(text)) !== null) {
      // Only add if not already detected with Sukoon
      const overlaps = spans.some(s => s.start === match!.index);
      if (!overlaps) {
        spans.push({
          start: match.index,
          end: match.index + match[0].length,
          rule: TajweedRule.QALQALAH,
          text: match[0],
        });
      }
    }

    return spans;
  }

  /**
   * Detect Madd (مَدّ) - Elongation
   */
  private detectMadd(text: string): TajweedSpan[] {
    const spans: TajweedSpan[] = [];

    const patterns = [
      new RegExp(`[${ALIF_MADD}]`, 'g'),                    // Alif Madd (آ)
      new RegExp(`[اوي]${MADD_SIGN}`, 'g'),                 // With Madd sign
      new RegExp(`[اوي][\\s]*${HAMZA_VARIANTS}`, 'g'), // Madd Munfasil
    ];

    patterns.forEach(pattern => {
      let match;
      while ((match = pattern.exec(text)) !== null) {
        spans.push({
          start: match.index,
          end: match.index + match[0].length,
          rule: TajweedRule.MADD,
          text: match[0],
        });
      }
    });

    return spans;
  }

  /**
   * Remove overlapping spans - keep higher priority rules
   */
  private removeOverlaps(spans: TajweedSpan[]): TajweedSpan[] {
    if (spans.length === 0) return spans;

    // Sort by start position
    spans.sort((a, b) => a.start - b.start);

    // Priority: Iqlab > Idghaam > Ghunnah > Ikhfa > Qalqalah > Madd
    const priorities: Record<TajweedRule, number> = {
      [TajweedRule.IQLAB]: 6,
      [TajweedRule.IDGHAAM]: 5,
      [TajweedRule.GHUNNAH]: 4,
      [TajweedRule.IKHFA]: 3,
      [TajweedRule.QALQALAH]: 2,
      [TajweedRule.MADD]: 1,
    };

    const result: TajweedSpan[] = [];
    let current: TajweedSpan | null = null;

    for (const span of spans) {
      if (!current) {
        current = span;
        continue;
      }

      // Check for overlap
      if (span.start < current.end) {
        // Overlapping - keep higher priority
        const currentPriority = priorities[current.rule] || 0;
        const spanPriority = priorities[span.rule] || 0;

        if (spanPriority > currentPriority) {
          current = span;
        }
      } else {
        // No overlap - add current and move to next
        result.push(current);
        current = span;
      }
    }

    if (current) {
      result.push(current);
    }

    return result;
  }
}
